### PR TITLE
feat: Add engine for CA EDD web

### PR DESCRIPTION
**Pull request by KevinJBoyer**
## Ticket

https://navalabs.atlassian.net/browse/DST-521

## Changes
 - Add engine that retrieves from the `ca-edd-web` dataset
 - Add Document.source as a field to track the URI
 - Add formatter for these
 - Split pages into chunks

## Context for reviewers
 - Deferred to a follow up PR: tracking Markdown headings and adding these to chunk.headings

### CODE REVIEW

File: app/src/ingest_edd_web.py [hunk 1]
```diff
@@ -32,18 +32,47 @@ def _ingest_edd_web(
         urls_processed.add(item["url"])
 
         content = item.get("main_content", item.get("main_primary"))
-        assert content, f"Item {name} has no main_content or main_primary"
+        if not content:
+            logger.warning("Item %s has no main_content or main_primary", name)
+            continue
```
Comment (yoomlam) [inline app/src/ingest_edd_web.py:37]: This should not occur. If it does, LMK which URL.

File: app/src/ingest_edd_web.py [hunk 2]
```diff
@@ -32,18 +32,47 @@ def _ingest_edd_web(
...
+    for subsection in content_split_by_double_newlines[1:]:
+        tokens_with_new_subsection = len(
+            embedding_model.tokenizer.tokenize(subsections[-1] + subsection)
+        )
+        if tokens_with_new_subsection < embedding_model.max_seq_length:
+            subsections[-1] += "\n\n" + subsection
+        else:
+            subsections.append(subsection)
+
+    # Parallelize embedding generation for performance
```
Comment (yoomlam) [inline app/src/ingest_edd_web.py:65]: Suggest adding a log message so we get a sense of the number of chunks per webpage.

File: app/src/format.py [hunk 3]
```diff
@@ -137,6 +137,62 @@ def format_bem_subsections(
...
+                    aria-expanded="false"
+                    aria-controls="a-{_accordion_id}">
+                    {citation_number}. {chunk.document.name}
+                </button>
+            </h4>
+            <div id="a-{_accordion_id}" class="usa-accordion__content usa-prose" hidden>
+                <div class="margin-left-2 border-left-1 border-base-lighter padding-left-2">{formatted_subsection}</div>
+                {citation_link}
+            </div>
+        </div>"""
```
Comment (yoomlam) [inline app/src/format.py:181]: Would be nice to refactor this -- https://github.com/navapbc/labs-decision-support-tool/pull/93/files#r1793952914

**Review by yoomlam — APPROVED** [review]

File: app/src/ingest_edd_web.py [hunk 4]
```diff
@@ -32,18 +32,47 @@ def _ingest_edd_web(
...
+def _chunk_page(content: str) -> list[Chunk]:
+    embedding_model = app_config.sentence_transformer
+
+    # Split content by double newlines, then gather into the largest chunks
+    # that tokenize to less than the max_seq_length
+    content_split_by_double_newlines = content.split("\n\n")
+    subsections = [content_split_by_double_newlines[0]]
+    for subsection in content_split_by_double_newlines[1:]:
+        tokens_with_new_subsection = len(
+            embedding_model.tokenizer.tokenize(subsections[-1] + subsection)
```
Comment (yoomlam) [inline app/src/ingest_edd_web.py:58]: Much cleaner code. For speed, I tried to reduce the number of calls to the embedding model by estimating the number of target chunks but the advantages is not yet clear.
https://github.com/navapbc/labs-decision-support-tool/blob/acadee85a8beee68d5aafbbab1bb90d1955f6c63/app/src/ingest_bem_pdfs.py#L201-L206

**Review by ccheng26 — APPROVED** [review]

File: app/src/ingest_edd_web.py (hunk 1 above)
Comment (KevinJBoyer) [reply app/src/ingest_edd_web.py:37]: Here's one:
> Processing EDD Online Application Now Available in Top Eight Languages  (https://edd.ca.gov/en/about_edd/news_releases_and_announcements/news-24-12/)

File: app/src/ingest_edd_web.py (hunk 4 above)
Comment (KevinJBoyer) [reply app/src/ingest_edd_web.py:58]: Oh right, thanks for reminding me about that code -- I wonder how expensive `embedding_model.tokenizer.tokenize` is

File: app/src/ingest_edd_web.py (hunk 2 above)
Comment (KevinJBoyer) [reply app/src/ingest_edd_web.py:65]: There's a log statement after this function is called, so the log statements look like
> 2024-10-10 19:13:08,728 - INFO - Processing Quick Statistics Data Definitions (https://edd.ca.gov/en/about_edd/quick_statistics_definitions/)
2024-10-10 19:13:13,322 - INFO - Split Quick Statistics Data Definitions into 6 chunks

File: app/src/format.py (hunk 3 above)
Comment (KevinJBoyer) [reply app/src/format.py:181]: I agree, will tackle in follow-up PR (to parse out headings as well)

File: app/src/ingest_edd_web.py (hunk 1 above)
Comment (yoomlam) [reply app/src/ingest_edd_web.py:37]: I have content for that:
```
"main_content": "[Español](/es/about_edd/news_releases_and_announcements/edd-online-application-now-available-in-top-eight-languages/)\n\n[Print](javascript:print();)\n\n# EDD Online Application Now Available in Top Eight Languages\n\nPublished: March 28, 2024\n\nNR No. 24-12  \nContact: Loree Levy/Aubrey Henry  \n916-654-9029  \n[mediainquiries@edd.ca.gov](mailto:mediainquiries@edd.ca.gov)\n\n**What You Need to Know:** California’s Armenian, Korean, and Tagalog communities can now apply online for unemployment benefits in their preferred language. This enhancement adds to English, Spanish, [Simplified Chinese, Traditional Chinese, and Vietnamese](https://edd.ca.gov/en/about_edd/news_releases_and_announcements/news-23-48/) languages already available. Together, these eight languages represent the top spoken languages of working-age adults in California.\n\n**SACRAMENTO**– Today, the Employment Development Department (EDD) announced ...
```
Retry the scraping from `main` -- you probably run the older code.

File: app/src/ingest_edd_web.py (hunk 4 above)
Comment (yoomlam) [reply app/src/ingest_edd_web.py:58]: If you use or adapt that code, start with a lower multiplication factor than `1.5`, like `1.2`.

File: app/src/ingest_edd_web.py (hunk 1 above)
Comment (KevinJBoyer) [reply app/src/ingest_edd_web.py:37]: You're right, rescraping fixed it. Will restore the assert!

**Pull request merged**